badger/attrcache.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2017 Pilosa Corp.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package badger
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+func newCache(s *attrStore) *attrCache {
+	return &attrCache{
+		attrStore: s,
+		attrs:     make(map[uint64]map[string]interface{}),
+		mu:        sync.RWMutex{},
+	}
+}
+
+//attrCache wraps the attrstore and provides a cache
+type attrCache struct {
+	//cachableStore
+	*attrStore
```

Comment:
In this case I decided to create a database layer and a caching layer:
* I wanted to test the database without the cache.
* It makes it a bit simpler to write benchmarks.
* It also seemed like it would be easy to reuse it in BoltDB.
* I am also not a great supporter of in memory maps as cache: 
  - I prefer a cache that release memory  
  - I remember reading that maps performance degrades after x millions of items. I figured this leaves things a bit more pliable.

However, without high cardinality data and real use cases these are all just preferences. It would be interesting to explore this further.